forc-pkg/src/source/reg/index_file.rs

```diff
@@ -0,0 +1,391 @@
+//! This module handles everything to do with index files.
+//!
+//! Index files are for creating set of information for identifying a published
+//! package. They are used by forc while fetching to actually convert a registry
+//! index into a IPFS CID. We also add some metadata to this index files to
+//! enable forc to do "more clever" fetching during build process. By moving
+//! dependency resolution from the time a package is fetched to the point we
+//! start fetching we are actively enabling forc to fetch packages and their
+//! dependencies in parallel.
+//!
+//! There are two main things forc needs to be able to do for index files:
+//!   1: Creation of index files from published packages
+//!   2: Calculating correct path for given package index.
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+
+#[derive(Serialize, Deserialize, Default)]
+pub struct IndexFile {
+    /// Each published instance for this specific package, keyed by their
+    /// versions. The reason we are doing this type of mapping is for use of
+    /// ease and deterministic ordering, we are effectively duplicating version
+    /// of package but keeping `PackageEntry` self contained.
+    #[serde(flatten)]
+    versions: BTreeMap<semver::Version, PackageEntry>,
+}
+
+/// A unique representation of each published package to `forc.pub`. Contains:
+///
+/// 1. The name of the package.
+/// 2. The version of the package.
+/// 3. CID of the package's source code. This is how forc actually resolves a
+///    package name, version information into actual information on how to get
+///    the package.
+/// 4. CID of the package's abi if the package is a contract.
+/// 5. Dependencies of this package. If there are other packages this package
+///    depends on, some information can be directly found in the root package
+///    to enable parallel fetching.
+#[derive(Serialize, Deserialize, Clone)]
+pub struct PackageEntry {
+    /// Name of the package.
+    /// This is the actual package name needed in forc.toml file to fetch this
+    /// package.
+    #[serde(alias = "package_name")]
+    name: String,
+    /// Version of the package.
+    /// This is the actual package version needed in forc.toml file to fetch
+    /// this package.
+    version: semver::Version,
+    /// IPFS CID of this specific package's source code. This is pinned by
+    /// forc.pub at the time of package publishing and thus will be
+    /// available all the time.
+    source_cid: String,
+    /// IPFS CID of this specific package's abi. This is pinned by
+    /// forc.pub at the time of package publishing and thus will be
+    /// available all the time if this exists in the first place, i.e the
+    /// package is a contract.
+    abi_cid: Option<String>,
+    /// Dependencies of the current package entry. Can be consumed to enable
+    /// parallel fetching by the consumers of this index, mainly forc.
+    dependencies: Vec<PackageDependencyIdentifier>,
+}
```

Comment:
nit: should we also add `yanked: bool` here?

Reply:
Yeah that's a good idea. Happy for this to be in a follow up PR

Reply:
Merging this one not to dismiss reviews adding it in the follow-up right away, thanks for catching this :) #7077